tools/buildsys/src/manifest.rs

```diff
@@ -358,16 +488,73 @@ impl ManifestInfo {
     }
 }
 
+/// For the "top-level manifest", i.e. the thing that `buildsys` is building, only
+/// `build-dependencies` are valid. This is because we would need all artifacts before the top-level
+/// manifest's `build.rs` runs. Once we go deeper in the graph, then both `build-dependencies` and
+/// `dependencies` are valid because they would be built in time for the top-level `build.rs`.
+fn is_valid_dep(top_manifest_name: &str, link: &PackageLink<'_>) -> bool {
+    let is_top_level_manifest = link.from().name() == top_manifest_name;
+    let is_deeper_level_manifest = !is_top_level_manifest;
+    is_deeper_level_manifest || link.build().is_present()
+}
+
+fn is_manifest_type(pkg_metadata: &PackageMetadata, manifest_type: BuildType) -> bool {
+    let metadata_table = pkg_metadata.metadata_table();
+    match manifest_type {
+        BuildType::Package => metadata_table.get("build-package").is_some(),
+        BuildType::Kit => metadata_table.get("build-kit").is_some(),
+        BuildType::Variant => metadata_table.get("build-variant").is_some(),
+    }
+}
+
+fn find_id(name: &str, graph: &PackageGraph, manifest_type: BuildType) -> Option<PackageId> {
+    for pkg_metadata in graph.packages() {
+        if is_manifest_type(&pkg_metadata, manifest_type) && pkg_metadata.name() == name {
+            return Some(pkg_metadata.id().to_owned());
+        }
+    }
+    None
+}
+
+/// Lists include the "top-level manifest", i.e. the thing that `buildsys` is being asked to build.
+/// We do not want this, we want only a list of things that it depends on. Here we convert
+/// `PackageMetadata` objects to the `String` name, and filter out the "top-level manifest".
+fn filter_map_to_name(top_manifest_name: &str, pkg_metadata: &PackageMetadata) -> Option<String> {
+    if pkg_metadata.name() == top_manifest_name {
+        None
+    } else {
+        // Return the package override name, if it exists, or else the Cargo manifest name.
+        Some(get_buildsys_package_name(pkg_metadata))
+    }
+}
```

Comment:
Are we guaranteed that the first entry in the traversal is the node we started from? Naively, that's what I would expect from a graph iterator. If so, we could just skip the first element from the iterator.

Reply:
> Are we guaranteed that the first entry in the traversal is the node we started from? Naively, that's what I would expect from a graph iterator. If so, we could just skip the first element from the iterator.

The documentation is silent on order of returned results. Seems safer and cheap to not make the assumption.